src/estimation/include/iDynTree/Estimation/ExtendedKalmanFilter.h

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright (C) 2018 Fondazione Istituto Italiano di Tecnologia
+ *
+ * Licensed under either the GNU Lesser General Public License v3.0 :
+ * https://www.gnu.org/licenses/lgpl-3.0.html
+ * or the GNU Lesser General Public License v2.1 :
+ * https://www.gnu.org/licenses/old-licenses/lgpl-2.1.html
+ * at your option.
+ */
+
+#ifndef EXTENDED_KALMAN_FILTER_H
+#define EXTENDED_KALMAN_FILTER_H
+
+#include <iDynTree/Core/VectorDynSize.h>
+#include <iDynTree/Core/MatrixDynSize.h>
+#include <iDynTree/Core/Utils.h>
+#include <iDynTree/Core/EigenHelpers.h>
+
+namespace iDynTree
+{
+
+    /**
+     * @class DiscreteExtendedKalmanFilter naive base class implementation of discrete EKF with additive Gaussian noise
+     *
+     * @warning This class is not stand-alone but can be used meaningfully only by classes deriving from it.
+     * The system propagation function f() and measurement model function h() are virtual functions to be defined by
+     * the derived class implementing the EKF.
+     * Similarly, the Jacobians for the linearized system propagation and measurement model are also virtual functions
+     * to be implemented by the derived class.
+     *
+     * The derived class must must set the size for states, inputs and outputs using ekfSetStateSize(), ekfSetInputSize() and ekfSetOutputSize() methods.
+     * Then the derived class must call ekfInit() to resize the vectors and matrices for the EKF.
+     *
+     * Similarly before running the estimator through a loop, it is recommended to set the initial states and variances for measurements, system dynamics and intial states.
+     * This is necessary to avoid any NaN values to be propagated.
+     *
+     * Once, initialized properly, the filter can be run by
+     *  - calling ekfSetInputVector() to set the control inputs and then calling ekfPredict() at each prediction step, and
+     *  - calling ekfSetMeasurementVector() to set the measurements and then calling ekfUpdate() at each update step
+     *
+     * The internal state of the estimator can be obtained by calling ekfGetStates()
+     */
+    class DiscreteExtendedKalmanFilter
+    {
+    protected:
+        DiscreteExtendedKalmanFilter();
+        virtual bool f(const iDynTree::VectorDynSize& x_k,
+                       const iDynTree::VectorDynSize& u_k,
+                       const iDynTree::VectorDynSize& w_k,
+                       iDynTree::VectorDynSize& xhat_k_plus_one) = 0;
+
+        virtual bool h(const iDynTree::VectorDynSize& xhat_k_plus_one,
+                       const iDynTree::VectorDynSize& v_k_plus_one,
+                       iDynTree::VectorDynSize& zhat_k_plus_one) = 0;
+
+        virtual bool computejacobianF(iDynTree::VectorDynSize& x, iDynTree::MatrixDynSize& F) = 0;
+        virtual bool computejacobianH(iDynTree::VectorDynSize& x, iDynTree::MatrixDynSize& H) = 0;
+
+        bool ekfPredict();
```

Comment:
~~The class is already called `DiscreteExtendedKalmanFilter`, perhaps it could make sense to drop the prefix in front of the methods? Or there is any specific reason to keep them?~~
I see, the class is meant to be overloaded, and the prefix is used to distinguish between `ekf` functions and child classes function, that could be named similarly. Given that this is the case, for the same reason it could make sense to also add a prefix to virtual function that need to be inherited?